galley/pkg/config/analysis/analyzers/namespaceconflict/authorizationpolicy.go

```diff
@@ -0,0 +1,114 @@
+package namespaceconflict
+
+import (
+	"fmt"
+
+	k8s_labels "k8s.io/apimachinery/pkg/labels"
+
+	v1beta1 "istio.io/api/security/v1beta1"
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/pkg/config/resource"
+	"istio.io/istio/pkg/config/schema/collection"
+	"istio.io/istio/pkg/config/schema/collections"
+)
+
+// Analyzer checks conditions related to conflicting namespace level resources
+type AuthorizationPolicyConflictAnalyzer struct{}
+
+var _ analysis.Analyzer = &AuthorizationPolicyConflictAnalyzer{}
+
+var authpolicyCol = collections.IstioSecurityV1Beta1Authorizationpolicies.Name()
+
+// Metadata implements Analyzer
+func (a *AuthorizationPolicyConflictAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name:        "namespaceconflict.AuthorizationPolicyConflictAnalyzer",
+		Description: "Checks conditions related to Authorization Policies conflicting namespace level resources",
+		Inputs: collection.Names{
+			authpolicyCol,
+		},
+	}
+}
+
+func (a *AuthorizationPolicyConflictAnalyzer) Analyze(c analysis.Context) {
+	namespaceWideConfiguration := map[string]*resource.Instance{}
+
+	c.ForEach(authpolicyCol, func(r *resource.Instance) bool {
+		x := r.Message.(*v1beta1.AuthorizationPolicy)
+		xNS := r.Metadata.FullName.Namespace.String()
+
+		// If the resource has workloads associated with it, analyze for conflicts with selector
+		if x.GetSelector() != nil {
+			// If there's namespace wide configuration, there's a conflict.
+			if val, found := namespaceWideConfiguration[xNS]; found {
+				conflicts := []string{val.Metadata.FullName.String(), r.Metadata.FullName.String()}
+				m := msg.NewNamespaceResourceConflict(r, authpolicyCol.String(), xNS, fmt.Sprintf("(ALL) Namespace: %v", xNS), conflicts)
+				c.Report(collections.IstioSecurityV1Beta1Peerauthentications.Name(), m)
```

Comment:
is this reporting the correct one as conflicting? Ideally if we have two resources, a and b, it should say:
`"Resource a conflicts with b and is ignored"` or similar.

Right now I am not sure the ordering is consistent with the internal logic (which, IIRC, uses the oldest one)

Reply:
Hey @howardjohn  (again) 😅. Right now I build the message based on what @esnible suggested in the issue, please do correct me if I got it wrong or if there's a better way to do it 🤔.


For example right now it outputs something like this for namespace wide configurations (I added the (ALL) to denote the superset of all resources):
```
Warning [IST0146] (PeerAuthentication defaultconfig2.default) More than one istio/security/v1beta1/peerauthentications in namespace "default" for workload "(ALL) Namespace: default": [default/defaultconfig default/defaultconfig2].
```

and this for specific selector conflicts in a given namespace, where [foo/selfinancefoo foo/selfinancefoo2] is the array of all resources associated with a same selector
 ```
Warning [IST0146] (PeerAuthentication selfinancefoo.foo) More than one istio/security/v1beta1/peerauthentications in namespace "foo" for workload "app=finance": [foo/selfinancefoo foo/selfinancefoo2].
 ```
 
 Where the message is defined with the following template:
 ```
   - name: "NamespaceResourceConflict"
    code: IST0146
    level: Warning
    description: "Multiple specifiers of the same kind in a namespace select the same workload."
    template: "More than one %s in namespace %q for workload %q: %v."
    args:
      - name: kind
        type: string
      - name: namespace
        type: string
      - name: workloadSelector
        type: string
      - name: conflictingCRs
        type: "[]string"
 ```
 
 Do you think this makes sense or should I remove the [] string and replace with specific "a is conflicting with b" kind of message? 
 
 Again, thanks a lot for looking at this so quickly ! :)